refactor: add service package and start splitting HTTP handling

ThinkChaos opened the pull request:
Progress for #1206.

No user visible changes in this PR: it's just refactoring and the new config cannot be set from the YAML file.  
Instead the new config structs are filled with the info from `ports:` until we're ready to deprecate `ports` fully.

The first part of the commits refactor how we setup the HTTP listeners.
The second part adds a new `service` package with interfaces used to split up `server.Server`.  
The commits after adding that package refactor the DoH, metrics, and API to be services.

Here's the docs that package as an overview of the design:
> Package service exposes types to abstract services from the networking.
> 
> The idea is that we build a set of services and a set of network
> endpoints (Listener). The services are then assigned to endpoints based
> on the address(es) they were configured for.
> 
> Actual service to endpoint binding is not handled by the abstractions in
> this package as it is protocol specific.
> The general pattern is to make a "server" that wraps a service, and can
> then be started on an endpoint using a `Serve` method,
> similar to `http.Server`.
> 
> To support exposing multiple compatible services on a single endpoint
> (example: DoH + metrics on a single port), services can implement `Merger`.

This doesn't touch DNS listener/server code (besides moving the TLS config to a single spot), but I think we'll be able to re-use the same code to setup DNS listeners, and split the DNS server logic into a bunch of services + one server.

The follow-up work is to continue removing things from `configureHTTPRouter` until it's empty and we can remove the transitional `httpMiscService`.  
At that point we'll be ready to deprecate `ports` and switch to the new `services` config format under discussion in #1206.

ThinkChaos commented:
I still need to write tests for the new package, and the various new "services", but besides that it's in a reviewable state :)

0xERR0R commented:
Skimmed over your PR and it looks very good! :+1:

kwitsch reviewed (approved):
Read through it and I think you did an amazing job. 👍

This approach looks like a really good foundation for a step wise implementation without end-user obstruction.

ThinkChaos commented:
I'm satisfied with where tests are at now. Codecov doesn't seem to understand the PR being based on a branch that's not targeting main (or maybe it's because I changed it), so there's no explicit percentage diff, but most new code is covered.  
The total coverage dropped 2% because I added a suite to the metrics package, but that's just more realistic :smile:

Also the e2e tests have been pretty flaky for me today, not sure if it's my changes or not.
Here's an example failure: https://github.com/0xERR0R/blocky/actions/runs/10639898980/job/29498704949

ThinkChaos commented:
@t-e-s-tweb thanks for the interesting and trying it out, there's no

ThinkChaos commented:
Then building the branch and making no changes to the config should work.  
The `CopyPortsToServices` function in `config.go` takes care of mapping the existing YAML structure to the new code.

ThinkChaos commented:
GitHub auto closed this with no option for me to reopen because the target branch was merged into main.
So I made a new PR: #1595.

(I was hoping GitHub had the same logic as GitLab where merging a MR changes the other MRs targeting that branch to now target where the first MR was merged, but no I guess  not)
